File: Python3/프로그래머스/2/42584. 주식가격/주식가격.py
def solution(prices):
    #가격이 떨어지지 않은 기간은 몇초인가?
    stack = []
    answer = [0 for _ in range(len(prices))]
    for i, p in enumerate(prices):
        if not stack:
            stack.append(i)
            continue
        
        if prices[stack[-1]] <= p:
            stack.append(i)
        else:
            #마지막 스택 값보다 작을 경우
            while stack:
                if prices[stack[-1]] > p:
                    l = stack.pop()
                    answer[l] = i - l
                else:
                    break
            stack.append(i)
    while stack:
        l = stack.pop()
        answer[l] = len(prices) - l - 1
    
    return answer
        
        
        
    
    
    
    
    
    
    
    
    
    
    
    
    
    
#     answer = [0] * len(prices)
#     stack = []
    
#     for i,p in enumerate(prices):
#         if not stack:
#             stack.append((i,p))
#             continue
        
#         if stack[-1][1] <= p:
#             stack.append((i,p))
#             continue
        
#         while stack and stack[-1][1] > p:
#             ti, tp = stack.pop()
#             answer[ti] = i-ti
#         stack.append((i,p))
    
#     while stack:
#         ti, tp = stack.pop()
#         answer[ti] = len(prices) - ti - 1
        
#     return answer

File: C++/프로그래머스/2/42584. 주식가격/주식가격.cpp
#include <string>
#include <vector>

using namespace std;
//가격이 떨어지지 않은 기간이 몇 초인가??
//스택에서 마지막 요소와 비교해서 계속 증가하는 경우 쌓기, 
vector<int> stack;

vector<int> solution(vector<int> prices) {
    
    int size = prices.size();
    vector<int> answer = vector<int>(size);
    int v = 0;
    for (int i = 0; i < size; i++) {
        if (stack.empty()) {
            stack.push_back(i);
            continue;
        }
        if (prices[stack.back()] <= prices[i]) {
            stack.push_back(i);
            continue;
            
        } else { //작다면, 가격이 떨어졌다면
            while (!stack.empty() && prices[stack.back()] > prices[i]) {
                v = stack.back();
                stack.pop_back();
                answer[v] = i - v;
            }
            stack.push_back(i);
        }
    }
    
    while (!stack.empty()) {
        v = stack.back();
        stack.pop_back();
        answer[v] = size - v - 1;
    }
    return answer;
}